Expose OT values to rules

CurlyMoo opened the pull request:
Todo:

- [x] Move values to PROGMEM

CurlyMoo commented:
I don't have the ability to test the changes so i'm curious how i did without it 😄

CurlyMoo commented on HeishaMon/HeishaOT.h:
```diff
@@ -1,3 +1,6 @@
+#ifndef _HEISHA_OT_H_
+#define _HEISHA_OT_H_
```
@IgorYbema Make this a good practice in all your header files. I prevent a lot of annoyance when including header files.

IgorYbema reviewed (commented):
Als ik het nu begrijp denk ik dat ook dit niet goed is.
Alleen READ en READWRITE values zijn door heishamon aan te sturen. 
WRITE values zijn waardes die de thermostaat stuurt naar Heishamon en dus NIET te bedienen zijn (en dus readonly voor heishamon). Zijn termen uit de OT wereld die het zo beschrijft ;-)

Ik denk dat de beste fix is om in de definitie van de values in heishaot.cpp de RW op 2 te zetten en WRITE op 3. Dan klopt deze aanpassing wel.

CurlyMoo commented:
@IgorYbema Ik heb simpelweg de conditie aangepast van `>=` naar `<=` en van `<=` naar `>=`. Dan werkt het ook.

IgorYbema commented:
> @IgorYbema Ik heb simpelweg de conditie aangepast van `>=` naar `<=` en van `<=` naar `>=`. Dan werkt het ook.

Nee want nu zijn de RW's (die 3 zijn) niet writable en de WRITE's (die readonly zijn, #2) wel writable nog :)

CurlyMoo commented:
Dan zou het nu moeten kloppen.

IgorYbema commented:
Top! Dan ga ik zelf nog even wat testen van de PR ergens vandaag

IgorYbema commented:
@CurlyMoo first analyse: crashes indeed when OT is connected (heatpump, s0, 1wire doesn't matter)

IgorYbema commented:
Fatal exception 3(LoadStoreErrorCause):
                                       epc1=0x40208881, epc2=0x00000000, epc3=0x00000000, excvaddr=0x4026d99c, depc=0x00000000

                                                                                                                              --------------- CUT HERE FOR EXCEPTION DECODER ---------------

Exception (3):
epc1=0x40208881 epc2=0x00000000 epc3=0x00000000 excvaddr=0x4026d99c depc=0x00000000

>>>stack>>>

ctx: cont
sp: 3ffffc50 end: 3fffffd0 offset: 0150
3ffffda0:  302e3236 00303030 07000000 401007dc
3ffffdb0:  00007fff 00000000 00000005 3ffffde0
3ffffdc0:  00000009 3ffffdd0 3fff3123 40205fcc
3ffffdd0:  302e3236 00303030 00000022 ffffffff
3ffffde0:  00007fff 00997812 3ffeed0c 401039b4
3ffffdf0:  3ffeb1f0 00000000 00000000 00000000
3ffffe00:  4026c0bb 00997812 40103e6e 00000100
3ffffe10:  3ffeb1f0 7fffffff 00002200 00000001
3ffffe20:  00000001 00004288 3fff16c8 00000030
3ffffe30:  3ffeb1f0 009952fc 00000001 3ffeeef0
3ffffe40:  4026c182 3fffff00 00000000 3fffdad0
3ffffe50:  3fff16c8 3fffc200 00000022 3fff16c8
3ffffe60:  3fffdad0 00000000 3fffff00 4021eab4
3ffffe70:  00000000 0000274e 3db22d0e 00000077
3ffffe80:  00000000 0000274e 820c49ba 0000004f
3ffffe90:  0000274e 3fff003c 00000030 4021ff5a
3ffffea0:  0000002a 3ffffed2 3ffefff0 402274b4
3ffffeb0:  0000002a 3ffffed0 00000003 3fffff28
3ffffec0:  0000002a 00000000 3ffefff0 402275df
3ffffed0:  0030002c 0000274e 283126e9 002d725e
3ffffee0:  3fff3117 3ffe9118 0000002b 002d7264
3ffffef0:  3fff1378 3ffefff0 3ffefff0 00000001
3fffff00:  00000001 0000274e 3ffefff0 40227b38
3fffff10:  00000007 3fff3139 3fff3117 4010123a
3fffff20:  3fff1378 00000000 3fff4501 4022a614
3fffff30:  00000003 00000021 0000002c 00000004
3fffff40:  3fffdad0 3fff1378 3fff169c 3fff16c8
3fffff50:  3fffdad0 00000000 3ffefff0 40207872
3fffff60:  00000000 3fff0031 00000000 00000000
3fffff70:  00e0018f 00000000 00000000 00ea018f
3fffff80:  00000000 feefeffe 00000000 402317a8
3fffff90:  3fff3744 feef0035 feefeffe 3fff16c8
3fffffa0:  3fffdad0 00000000 3fff169c 3fff16c8
3fffffb0:  3fffdad0 00000000 3fff169c 4022e1e4
3fffffc0:  feefeffe feefeffe 3fffdab0 401019b5
<<<stack<<<

--------------- CUT HERE FOR EXCEPTION DECODER ---------------


Exception 3: LoadStoreError: Processor internal physical address or data error during load or store
PC: 0x40208881
EXCVADDR: 0x4026d99c

Decoding stack results
0x401007dc: ets_post(uint8, ETSSignal, ETSParam) at \Arduino15\packages\esp8266\hardware\esp8266\3.1.2\cores\esp8266\core_esp8266_main.cpp line 238
0x40205fcc: mqtt_callback(char*, unsigned char*, unsigned int) at \HeishaMon\HeishaMon/HeishaMon.ino line 505
0x4021eab4: ESP8266WiFiAPClass::softAPSSID() const at \Arduino15\packages\esp8266\hardware\esp8266\3.1.2\libraries\ESP8266WiFi\src\ESP8266WiFiAP.cpp line 361
0x4021ff5a: WiFiClient::read() at \Arduino15\packages\esp8266\hardware\esp8266\3.1.2\libraries\ESP8266WiFi\src/include/ClientContext.h line 249
0x402274b4: PubSubClient::readByte(unsigned char*) at C:\Users\igor\Documents\Arduino\libraries\PubSubClient\src\PubSubClient.cpp line 297
0x402275df: PubSubClient::readPacket(unsigned char*) at C:\Users\igor\Documents\Arduino\libraries\PubSubClient\src\PubSubClient.cpp line 350
0x40227b38: PubSubClient::loop() at \arduino15\packages\esp8266\tools\xtensa-lx106-elf-gcc\3.1.0-gcc10.3-e5f9fec\xtensa-lx106-elf\include\c++\10.3.0\bits/std_function.h line 617
0x4010123a: umm_free_core(umm_heap_context_t*, void*) at \Arduino15\packages\esp8266\hardware\esp8266\3.1.2\cores\esp8266\umm_malloc\umm_malloc.cpp line 642
0x4022a614: HardwareSerial::available() at \Arduino15\packages\esp8266\hardware\esp8266\3.1.2\cores\esp8266\HardwareSerial.cpp line 111
0x40207872: loop() at \HeishaMon\HeishaMon/HeishaMon.ino line 1154
0x4022e1e4: loop_wrapper() at \Arduino15\packages\esp8266\hardware\esp8266\3.1.2\cores\esp8266\core_esp8266_main.cpp line 258

IgorYbema commented:
Gaat mis net na mijn mqtt callback:
Sat Apr  1 07:29:12 2023 (10063): OpenTherm: MQTT message received 'dhwSetpoint'

Volgens stacktrace ook in de callback.

IgorYbema commented:
Ok nailed it down to these both calls:
getOTStructMember(F("outsideTemp"))->value.f = String(value).toFloat();
rules_event_cb(F("?"), topic);
Probably for other recalled values also.
Both create a crash, doesn't matter which line I remove.

CurlyMoo commented:
Gaan we verder in Nederlands of Engels? Voor nu maar even in het Nederlands. Op een of andere manier gaat het dan dus mis bij de PROGMEM `F` functie. Even kijken wat ik daar dan anders doe dan elders, waar het wel goed gaat.

CurlyMoo commented:
Kan je kijken waar precies in `getOTStructMember` het dan mis gaat. Daar zal het ergens moeten zitten. In die functie zet ik nu `name` om naar een reguliere `str`, maar ik heb het vermoeden dat we ook `heishaOTDataStruct[i].name` netjes uit PROGMEM moeten halen voordat we `strcmp_p` aanroepen. Maar nu ik weet waar het fout gaat, kan ik dit even losstaand unittesten.

IgorYbema commented:
Gaat inderdaad mis sinds de progmem commit. Net even die uit mijn test er uit gehaald en het draait weer.

CurlyMoo commented:
Zoveel was duidelijk na de feedback op tweakers.

CurlyMoo commented:
Daarom heb ik de PROGMEM zooi bewaart tot het laatst. Dat is het lastigste uit het hoofd te doen. Dat blijkt wel weer.

IgorYbema commented:
Zou het kunnen dat:
`struct heishaOTDataStruct_t heishaOTDataStruct[] PROGMEM = {`
alleen de pointer in progmem zet en niet de hele struct?

IgorYbema commented:
Van mij hoeft die hele struct niet in progmem hoor. Zoveel bytes zijn dat er nou ook weer niet.

IgorYbema commented:
En anders laten we de progmem commit voor wat het is. Buiten deze struct is het verder ook niet spannend wat het zou moeten besparen in geheugen.

CurlyMoo commented:
Op zich best wel wat, want er zitten ook een hoop logs bij die niet in PROGMEM zaten.

IgorYbema commented:
Valt wel mee toch als ik de commit bekijk? Het zijn niet echt grote lompe data blokken

CurlyMoo commented:
Komt goed...

CurlyMoo commented:
Die struct moet inderdaad wel uit PROGMEM en niet om de ruimte die hij inneemt, maar omdat hij ook de waardes bevat van OT. Die moeten dus te schrijven zijn nadat de boel draait. Dat kan niet in PROGMEM. Enige optie is struct in tweeën splitsen, maar dat zou jammer zijn inderdaad.

Blijft dan alleen ` rules_event_cb(F("?"), topic);` nog over?

CurlyMoo commented:
Oftewel, PROGMEM weg bij de struct initializatie en de betreffende functie als volgt:
```c
struct heishaOTDataStruct_t *getOTStructMember(const __FlashStringHelper *name) {
  PGM_P p = (PGM_P)name;
  int len = strlen_P((const char *)p);
  char *str = (char *)MALLOC(len + 1);
  if (str == NULL) {
    OUT_OF_MEMORY
  }
  strcpy_P(str, p);

  int i = 0;
  while(heishaOTDataStruct[i].name != NULL) {
    if(strcmp(str, heishaOTDataStruct[i].name) == 0) {
      return &heishaOTDataStruct[i];
    }
    i++;
  }
  FREE(str);

  return NULL;
}
```

IgorYbema commented:
Ik ga het later vanavond of morgen eens testen daarmee. Krijg zo bezoek dus even geen tijd voor hobby :)

IgorYbema commented:
foutje met knop :)

CurlyMoo commented:
De laatste fix lost nog niet alle problemen op. Laat je de nieuwe stackdump weten?

IgorYbema commented:
Geen stackdump want een hardware watchdog reset tijdens 'reading rules'. Maw, het duurt te lang. Zit ergens in een loop.

CurlyMoo commented:
Is dat op basis van de laatste versie? De vastlopende loop was voorheen deze functie, wat dus blijkbaar niet zo werkt in C++
https://github.com/IgorYbema/HeishaMon/pull/112/commits/f250465f2f57d43c49a470b20faf5643ca4d68b0#diff-01e611e223b74b3014fcb8ea94fa38cb3fcb14850967c93186fc5dab1e1d1910L1401-L1404

Als het nu nog in een loop blijft hangen dan lijkt het erop dat hij ook bij de andere `rules_event_cb` recursief blijft aanroepen.

CurlyMoo commented:
Ik heb simpelweg even als unittest de volgende functie aanroepen in de `setup()` gezet en dat gaat prima:
```c
  rules_event_cb(F("?"), F("chEnable"));
  rules_event_cb(F("?"), "chEnable");
  rules_event_cb("?", "chEnable");
```

IgorYbema commented:
Is helaas nog niet stabiel. Mijn module hangt alleen aan opentherm en dan knalt hij er al uit na een paar minuten. Momenteel geen tijd om verder te debuggen. Webinterface is ook veel te traag nu

IgorYbema commented:
Wat mij betreft rollback tot vlak voor de progmem wijzigingen en de progmem dan in een nieuwe PR want die maakt het blijkbaar stuk nog

CurlyMoo commented:
@IgorYbema 

Previously i did this:
```c
char *progmem2char(const __FlashStringHelper *progmem) {
  PGM_P p = (PGM_P)progmem;
  int len = strlen_P((const char *)p);
  char *str = (char *)MALLOC(len + 1);
  if (str == NULL) {
    OUT_OF_MEMORY
  }
  strcpy_P(str, p);

  return str;
}

void log_message(const __FlashStringHelper *msg) {
  char *str = progmem2char(msg);

  log_message(str);

  FREE(str);
}

log_message(F("foo bar"));
```

And even before this:
```c
void log_message(const __FlashStringHelper *msg) {
  PGM_P p = (PGM_P)msg;
  int len = strlen_P((const char *)p);
  char *str = (char *)MALLOC(len + 1);
  if (str == NULL) {
    OUT_OF_MEMORY
  }
  strcpy_P(str, p);

  log_message(str);

  FREE(str);
}
```
Where the nested `log_message` calls the `char *` variant of the function.

But somehow, passing `const __FlashStringHelper *` as function parameter doesn't hold well. Therefor i changed this to:
```c
log_message(_F("foo bar"));
```

Which translates to:
```c
log_message((char *)(String(F("foo bar").c_str()));
```
Therefor translating the PROGMEM pointer to a regular `char *` before it is passed through as a function parameter. This removes all hassle with dealing with PROGMEM inside our functions. That's why i also removed that other `log_message` variant.

IgorYbema commented:
I will merge and fix some current merge issue's due to the new data block for K/L
